File: basic/src/aps.f90
!////////////////////////////////////////////////////!
! * Angular power spectrum calculation
!////////////////////////////////////////////////////!

module aps
  use constants, only: pi
  use pstool, only: binned_ells
  implicit none

  private pi
  private binned_ells

contains


subroutine binning(bn,eL,bp,bc,spc)
!* Return multipole-bin edges and centers
!* 
!* Args:
!*   :bn (int)    : number of bins
!*   :eL[2] (int) : bin edges
!*
!* Args(optional):
!*   :spc (str)   : bin spacing, '' = linear (default), 'log' = log spacing, 'log10' = log10 spacing, 'p2' = power of 2 spacing, 'p3' = power of 3 spacing
!*
!* Returns:
!*   :bp (double) : bin edges, with bounds (0:bn)
!*   :bc (double) : bin centers, with bounds (bn)
!*
  implicit none
  !I/O
  character(4), intent(in) :: spc
  integer, intent(in) :: bn
  integer, intent(in), dimension(2) :: eL
  double precision, intent(out), dimension(0:bn) :: bp
  double precision, intent(out), dimension(bn) :: bc
  !optional
  !opt4py :: spc = ''

  call binned_ells(eL,bp,bc,spc=spc)

end subroutine binning


subroutine map_vars(lmax,cl,sigma)
!*  Variance of a map and its derivative
!* 
!*  Args:
!*    :lmax (int)    : Maximum multipole of the input cl
!*    :cl[l] (double): Angular power spectrum, with bounds (0:lmax)
!*
!*  Returns:
!*    :sigma[2] (double): Variance of the map and of map derivative, with bounds (2)
!*  
  implicit none
  !I/O
  integer, intent(in) :: lmax
  double precision, intent(in), dimension(0:lmax) :: cl
  double precision, intent(out), dimension(0:1) :: sigma
  !internal
  integer :: l
  double precision :: al
  
  sigma = 0d0
  do l = 1, lmax
    al = dble(l)
    sigma(0) = sigma(0) + (2d0*al+1d0)*cl(l)
    sigma(1) = sigma(1) + (2d0*al+1d0)*(al**2+al)*cl(l)
  end do 
  sigma = sigma/(4d0*pi)

end subroutine map_vars


end module aps

File: wrap_py2/basic/aps.py
import libbasic

def binning(bn,eL,spc= ''):
  """
 Return multipole-bin edges and centers
 
 Args:
    :bn (*int*): number of bins
    :eL[*2*] (*int*): bin edges

 Args(optional):
    :spc (*str*): bin spacing, '' = linear (default), 'log' = log spacing, 'log10' = log10 spacing, 'p2' = power of 2 spacing, 'p3' = power of 3 spacing

 Returns:
    :bp (*double*): bin edges, with bounds (0:bn)
    :bc (*double*): bin centers, with bounds (bn)

  Usage:
    :bc,bp = basic.aps.binning(bn,eL,spc):
  """
  return libbasic.aps.binning(bn,eL,spc)

def read_cambcls(f,lmin,lmax,numcls,bb= 0,raw= 0):
  """
  Return CMB cls from CAMB output files

  Args:
    :f (*str*): Filename
    :lmin (*int*): Minimum multipole of the output Cl
    :lmax (*int*): Maximum multipole of the output Cl
    :numcls (*int*): Number of cls to be read

  Args(Optional):
    :bb (*bool*): Including BB in the file or not. The data should be TT, EE, TE, dd, Td (,Ed) if bb = False (default), and TT, EE, BB, TE if bb = True.
    :raw (*bool*): The cls in the file are multiplied by l(l+1)/2pi if raw = False (default)

  Returns:
    :cl[*numcls,l*] (*double*): Angular power spectra, with bountd (numcls,0:lmax)

  Usage:
    :cl = basic.aps.read_cambcls(f,lmin,lmax,numcls,bb,raw):
  """
  return libbasic.aps.read_cambcls(f,lmin,lmax,numcls,bb,raw)

def map_vars(lmax,cl):
  """
  Variance of a map and its derivative
 
  Args:
    :lmax (*int*): Maximum multipole of the input cl
    :cl[*l*] (*double*): Angular power spectrum, with bounds (0:lmax)

  Returns:
    :sigma[*2*] (*double*): Variance of the map and of map derivative, with bounds (2)
  
  Usage:
    :sigma = basic.aps.map_vars(lmax,cl):
  """
  return libbasic.aps.map_vars(lmax,cl)

def cl2bcl(bn,lmax,cl,spc= ''):
  """
  From unbinned to binned angular power spectrum

  Args:
    :bn (*int*): number of multipole bins
    :lmax (*int*): maximum multipole of the input angular power spectrum
    :cl[*l*] (*double*): angular power spectrum, with bounds (0:lmax)

 Args(optional):
    :spc (*str*): bin spacing, '' = linear (default), 'log' = log spacing, 'log10' = log10 spacing, 'p2' = power of 2 spacing, 'p3' = power of 3 spacing

  Returns:
    :cb[*bin*] (*double*): auto or cross angular power spectrum with multipole binning, with bounds (0:bn-1)

  Usage:
    :cb = basic.aps.cl2bcl(bn,lmax,cl,spc):
  """
  return libbasic.aps.cl2bcl(bn,lmax,cl,spc)

